File: android/src/main/java/com/wortise/ads/react/RNWortiseInterstitial.java
package com.wortise.ads.react;

import android.app.Activity;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import com.wortise.ads.AdError;
import com.wortise.ads.interstitial.InterstitialAd;

public class RNWortiseInterstitial extends ReactContextBaseJavaModule implements InterstitialAd.Listener, LifecycleEventListener {

  public static final String EVENT_CLICKED        = "onInterstitialClicked";
  public static final String EVENT_DISMISSED      = "onInterstitialDismissed";
  public static final String EVENT_FAILED_TO_LOAD = "onInterstitialFailedToLoad";
  public static final String EVENT_FAILED_TO_SHOW = "onInterstitialFailedToShow";
  public static final String EVENT_IMPRESSION     = "onInterstitialImpression";
  public static final String EVENT_LOADED         = "onInterstitialLoaded";
  public static final String EVENT_SHOWN          = "onInterstitialShown";


  private InterstitialAd mInterstitialAd;


  public RNWortiseInterstitial(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  @Override
  public String getName() {
    return "RNWortiseInterstitial";
  }

  @ReactMethod
  public void addListener(String eventName) {
  }

  @ReactMethod
  public void destroy() {
    if (mInterstitialAd == null) {
      return;
    }

    mInterstitialAd.destroy();
    mInterstitialAd = null;
  }

  @ReactMethod
  public void isAvailable(Promise promise) {
    promise.resolve((mInterstitialAd != null) && mInterstitialAd.isAvailable());
  }

  @ReactMethod
  public void isShowing(Promise promise) {
    promise.resolve((mInterstitialAd != null) && mInterstitialAd.isShowing());
  }

  @ReactMethod
  public void loadAd() {
    if (mInterstitialAd == null) {
      return;
    }

    mInterstitialAd.loadAd();
  }

  @ReactMethod
  public void removeListeners(Integer count) {
  }

  @ReactMethod
  public void setAdUnitId(String adUnitId) {
    Activity currentActivity = getCurrentActivity();

    if (currentActivity == null) {
        return;
    }

    destroy();

    mInterstitialAd = new InterstitialAd(currentActivity, adUnitId);
    mInterstitialAd.setListener(this);
  }

  @ReactMethod
  public void showAd(Promise promise) {
    Activity currentActivity = getCurrentActivity();

    if (mInterstitialAd == null) {
        promise.resolve(false);
        return;
    }

    if (currentActivity != null) {
      mInterstitialAd.showAd(currentActivity);
    } else {
      mInterstitialAd.showAd();
    }

    promise.resolve(true);
  }


  private void sendEvent(String eventName, WritableMap params) {
    getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(eventName, params);
  }


  @Override
  public void onHostDestroy() {
    destroy();
  }

  @Override
  public void onHostPause() {
  }

  @Override
  public void onHostResume() {
  }

  @Override
  public void onInterstitialClicked(InterstitialAd ad) {
    sendEvent(EVENT_CLICKED, null);
  }

  @Override
  public void onInterstitialDismissed(InterstitialAd ad) {
    sendEvent(EVENT_DISMISSED, null);
  }

  @Override
  public void onInterstitialFailedToLoad(InterstitialAd ad, AdError error) {
    WritableMap event = Arguments.createMap();

    event.putString("message", error.toString());
    event.putString("name",    error.name());

    sendEvent(EVENT_FAILED_TO_LOAD, event);
  }

  @Override
  public void onInterstitialFailedToShow(InterstitialAd ad, AdError error) {
    WritableMap event = Arguments.createMap();

    event.putString("message", error.toString());
    event.putString("name",    error.name());

    sendEvent(EVENT_FAILED_TO_SHOW, event);
  }

  @Override
  public void onInterstitialImpression(InterstitialAd ad) {
    sendEvent(EVENT_IMPRESSION, null);
  }

  @Override
  public void onInterstitialLoaded(InterstitialAd ad) {
    sendEvent(EVENT_LOADED, null);
  }

  @Override
  public void onInterstitialShown(InterstitialAd ad) {
    sendEvent(EVENT_SHOWN, null);
  }
}

File: android/src/main/java/com/wortise/ads/react/RNWortiseInterstitial.kt
package com.wortise.ads.react

import android.app.Activity
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.wortise.ads.AdError
import com.wortise.ads.RevenueData
import com.wortise.ads.interstitial.InterstitialAd
import com.wortise.ads.react.extensions.toWritableMap

class RNWortiseInterstitial(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext), InterstitialAd.Listener, LifecycleEventListener {

  private val eventEmitter by lazy {
    reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
  }

  private var interstitialAd: InterstitialAd? = null


  override fun getName(): String = "RNWortiseInterstitial"

  @ReactMethod
  fun addListener(eventName: String) {}

  @ReactMethod
  fun destroy() {
    interstitialAd?.destroy()
    interstitialAd = null
  }

  @ReactMethod
  fun isAvailable(promise: Promise) {
    promise.resolve(interstitialAd?.isAvailable == true)
  }

  @ReactMethod
  fun isShowing(promise: Promise) {
    promise.resolve(interstitialAd?.isShowing == true)
  }

  @ReactMethod
  fun loadAd() {
    interstitialAd?.loadAd()
  }

  @ReactMethod
  fun removeListeners(count: Int) {}

  @ReactMethod
  fun setAdUnitId(adUnitId: String) {
    val currentActivity = reactApplicationContext.currentActivity ?: return

    destroy()

    interstitialAd = InterstitialAd(currentActivity, adUnitId).also {
      it.listener = this
    }
  }

  @ReactMethod
  fun showAd(promise: Promise) {
    val interstitialAd = interstitialAd

    if (interstitialAd == null) {
      promise.resolve(false)
      return
    }

    val currentActivity = reactApplicationContext.currentActivity

    currentActivity?.let(interstitialAd::showAd) ?: interstitialAd.showAd()

    promise.resolve(true)
  }


  private fun sendEvent(eventName: String, params: WritableMap? = null) {
    eventEmitter.emit(eventName, params)
  }


  override fun onHostDestroy() {
    destroy()
  }

  override fun onHostPause() {}

  override fun onHostResume() {}

  override fun onInterstitialClicked(ad: InterstitialAd) {
    sendEvent(EVENT_CLICKED)
  }

  override fun onInterstitialDismissed(ad: InterstitialAd) {
    sendEvent(EVENT_DISMISSED)
  }

  override fun onInterstitialFailedToLoad(ad: InterstitialAd, error: AdError) {
    sendEvent(EVENT_FAILED_TO_LOAD, error.toWritableMap())
  }

  override fun onInterstitialFailedToShow(ad: InterstitialAd, error: AdError) {
    sendEvent(EVENT_FAILED_TO_SHOW, error.toWritableMap())
  }

  override fun onInterstitialImpression(ad: InterstitialAd) {
    sendEvent(EVENT_IMPRESSION)
  }

  override fun onInterstitialLoaded(ad: InterstitialAd) {
    sendEvent(EVENT_LOADED)
  }

  override fun onInterstitialRevenuePaid(ad: InterstitialAd, data: RevenueData) {
    sendEvent(EVENT_REVENUE_PAID, data.toWritableMap())
  }

  override fun onInterstitialShown(ad: InterstitialAd) {
    sendEvent(EVENT_SHOWN)
  }


  companion object {
    const val EVENT_CLICKED        = "onInterstitialClicked"
    const val EVENT_DISMISSED      = "onInterstitialDismissed"
    const val EVENT_FAILED_TO_LOAD = "onInterstitialFailedToLoad"
    const val EVENT_FAILED_TO_SHOW = "onInterstitialFailedToShow"
    const val EVENT_IMPRESSION     = "onInterstitialImpression"
    const val EVENT_LOADED         = "onInterstitialLoaded"
    const val EVENT_REVENUE_PAID   = "onInterstitialRevenuePaid"
    const val EVENT_SHOWN          = "onInterstitialShown"
  }
}
